ci: merge staging to master

maxwell-aiden opened the pull request:
This is an automatic PR generated by the pipeline CI/CD. This will be automatically fast-forward merged if successful.

@tegefaulkes @emmacasolin some todos left here:

9. [x] switch to staging as default branch - @CMCDragonkai 
12. [x] fix up the `build:mac` job to evaluate the homebrew shell environment - @tegefaulkes 
1. [x] use the new logger features to help debugging - @emmacasolin 
2. [x] merge the nat testing PR (without testnet.polykey.io testing) - @emmacasolin 
4. ~[ ] test load balancing and fixing up coverage reports again in the CI/CD~ - @emmacasolin has already made some progress on this - to be done in https://github.com/MatrixAI/TypeScript-Demo-Lib/issues/58 and https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/65 (new PR to be created by @emmacasolin)
5. [x] fixing up the integration build job involving `pkg.js`, use the debug log and find out why pkg is failing to package PK - @tegefaulkes 
   - Turns out a problem `tslib`, later issue to investigate and work out integration of `esbuild` or similar bundler
   - This leads to 2 additional issues:
     * https://github.com/MatrixAI/js-polykey/issues/389
     * https://github.com/MatrixAI/Polykey-CLI/issues/23
11. ~[ ] add in integration tests for windows/macos/ubuntu~ - to be done in MatrixAI/Polykey-CLI#10
    - @tegefaulkes (get your nixos updated and get tailscale installed, so I can do this https://tailscale.com/kb/1084/sharing/ and allow you to do quick testing on mac & windows)
    - `leveldown` is causing segfaults, but it's being replaced in js-db
3. [ ] benching and metrics (first get TS-Demo-Lib's benching to produce bench metrics that gitlab CI/CD understands and can visualise) - @emmacasolin 
6. [X] ensure that integration:merge job works, so the bot is actually merging this staging into master - @tegefaulkes 
    - [x] fix `integration:nix` job. The shebangs for `bin/polykey.js` was not getting patched during the `nix-build` process. The fix was to make `bin/polykey.ts` and `bin/polykey-agent.ts` executable.
    - [x] fix `integration:docker`. This was the same problem as above. 
8. ~[ ] deployment jobs - AWS - testnet deployment needs me involved~ - @tegefaulkes - to be done in MatrixAI/Polykey#194 
9. ~[ ] release jobs - check prerelease and release jobs are working by pushing up a tag with `git push --follow-tags`~ - @tegefaulkes - to be done in MatrixAI/Polykey#390 
10. [ ] priority switches to concurrency testing once I finish up DBTransaction for SI - @CMCDragonkai 

As long as we can deploy to testnet, then we want to merge this into master branch, so we can start new feature branches from the new features.

Once merged to master branch, it's time to do rename this to `Polykey` to match https://github.com/MatrixAI/Polykey/issues/5#issuecomment-1123140985. Then to update the CI/CD mirror repository links and integrate codesee.

Some side-issues to be addresses if they help us achieve the primary goals above:

1. Tracing context to help asynchronous and distributed debugging - https://github.com/MatrixAI/js-logger/issues/3 (we must have an adequate way of visualising the trace though for it to be useful)
2. Using `ts-node` with `swc` to speed up our `ts-node` executions and remove the typescript node cache
3. Using `esbuild` to workaround issues with using `pkg` as a bundler
4. Using `npm link` on our native addons to enable debug builds when debugging
5. And when will the SI be ready!?!?!?

joshuakarp committed:
test: NAT Traversal testing utils and tests

joshuakarp committed:
feat: added `nodesGetAll` bin command and service handler

Retrieves all buckets from the NodeGraph

CMCDragonkai committed:
fix: general fixes

- added getBuckets test for distance and lastUpdated order
- resetting buckets work
- changing utility names

tegefaulkes committed:
feat: added `NodeGraph.getClosestNodes()`

Implemented `getClosestNodes()` and relevant tests in `NodeGraph.test.ts`.

Relates to #212

tegefaulkes committed:
fix: `NodeManager.setNode` Properly handles adding new node when bucket is full

Logic of adding nodes has been split between `NodeManager` and `NodeGraph`. The `NodeGraph.setNode` just handles adding a node to the bucket where the `NodeManager.setNode` contains the logic of when to add the node

Relates #359

tegefaulkes committed:
test: fixed `nodeGraph` `get all buckets` test

fix into 7cc74d059f94739591307d9543fa73f999d3345f

tegefaulkes committed:
feat: added `connectionEstablishedCallback` to `Proxy`

Added a callback to the `Proxy` that is called when a `ForwardConnection` or `ReverseConnection` is established and authenticated. It is called with the following connection information; `remoteNodeId`, `remoteHost`, `remotePort` and `type`. They type signifies if it was a forward or reverse connection. Note that this is only triggered by composed connections.

Added a test for if the callback was called when a `ReverseConnection` is established.

Relates #332
Relates #344

tegefaulkes committed:
feat: `Proxy` trigger adding nodes to `Nodegraph`

Added an event to the `EventBus` that is triggered by the `Proxy`'s `connectionEstablishedCallback`. this adds the node to the `NodeGraph`.

Related #344

tegefaulkes committed:
feat: added optional timeout timers to `NodeConnectionManager` methods

In some cases we want to specify how long we attempt to connect to a node on a per-connection basis.

Related #363

tegefaulkes committed:
refactor: updated implementation of `nodePing`

Added `nodePing` command to `NodeConnectionManager` and `NodeManager.nodePing` calls that now. the new `nodePing` command essentially attempts to create a connection to the target node. In this process we authenticate the connection and check that the nodeIds match. If no address is provided it will default to trying to find the node through kademlia.

Related #322

CMCDragonkai committed:
fix: adding dependencies for `async-locks`, `db`, `resources` and `errors`

Updating dependencies to bring in new libraries such as locks, resources and errors. Some domains have been updated to use this change.

tegefaulkes committed:
feat: `NodeManager.setNode` authenticates the added node

`setNode` now authenticates the node you are trying to add. Added a flag for skipping this authentication as well as a timeout timer for the authentication. this is shared between authentication new node and the old node if the bucket is full.

Related #322

tegefaulkes committed:
syntax: general linting and fixes from api changes

tegefaulkes committed:
fix: squash into ping node changes.

Updated `NodeConnectionManager.pingNode` to just use the proxy connection.

#322

tegefaulkes committed:
fix: setNode now does not ping the new node

#322

tegefaulkes committed:
feat: setNode concurrently pings multiple nodes

`setNode` now pings 3 nodes concurrently, updating ones that respond and removing ones that don't. If there is room in the bucket afterwards then we add the new node.

#322

tegefaulkes committed:
feat: nodeManager is now startStop

#322

tegefaulkes committed:
feat: async queueing for setting nodes

`setNode` now has a `blocking` flag that defaults to false. If it encounters a full bucket when adding a node then it will add the operation to the queue and asynchronously trys a blocking `setNode` in the background. `setNode`s will only be added to the queue if the bucket was full.

#322

tegefaulkes committed:
feat: establishing a `NodeConnection` adds the node to the nodeGraph

`NodeConnectionManager` now takes `NodeGraph` in  the `nodeConnectionManager.start` method. It has to be part of the start method since they are co-dependent. `NodeConnectionManager` cals `NodeManager.setNode()` when a connection is established. This fulfills the condition of adding a node to the graph during a forward connection.

Fixed up tests that were failing in relation to the `NodeManager` `StartStop` conversion.

#322

tegefaulkes committed:
feat: implemented `NodeManager.refreshBucket()`

This method preforms the kademlia `refreshBucket` operation. It selects a random node within the bucket and preforms a search for that node. The process exchanges node information with any nodes it connects to.

#345

tegefaulkes committed:
feat: implemented no activity timers and queuing for `refreshBucket`

Added queuing for `refreshBucket`. This means that buckets will be refreshed one at a time sequentially. This is to avoid doing a lot of costly refreshing all at once.

Added no activity for buckets. If a bucket hasn't been touched for a while, 1 hour by default, it will add a refresh bucket operation to the queue. Timers are disabled for buckets already in the queue. Only 1 timer is used for all buckets since only one of them can have the shortest timer and that's all we really care about.

#345

tegefaulkes committed:
feat: refreshing buckets when entering network

`nodeConnectionManager.syncNodeGraph` now refreshes all buckets above the closest node as per the kademlia spec. This means adding a lot of buckets to the refresh bucket queue when an agent is started.

#345

tegefaulkes committed:
feat: abort controller support for `NodeManager.refreshBucket`

Added support to cancel out of a `refreshBucket` operation. This is to allow faster stopping of the `NodeManager` by aborting out of a slow `refreshBucket` operation. This has been implemented with the `AbortController`/`AbortSignal` API. This is not fully supported by Node14 so we're using the `node-abort-controller` to provide functionality for now.

#345

emmacasolin committed:
fix: `setNode` no longer adds node twice when `force` is true
#322

emmacasolin committed:
feat: generic `SetNodeQueue` class for queuing `setNode` operations
`NodeManager.setNode` and `NodeConnectionManager.syncNodeGraph` now utilise a single, shared queue to asynchronously add nodes to the node graph without blocking the main loop. These methods are both blocking by default but can be made non-blocking by setting the `block` parameter to false.
#322

emmacasolin committed:
fix: `syncNodeGraph` during agent startup is now non-blocking
#322

emmacasolin committed:
refactor: renamed `SetNodeQueue` to `Queue`

Renamed `queueStart` and `queuePush` since `Queue` is its own class now
Simplified some logic using the `promise` utility

emmacasolin committed:
build: upgrade to Node v16

Updated pkgs.nix to a5774e76bb8c3145eac524be62375c937143b80c
Updated node2nix to 1.11.0 and loading directly from GitHub
Updated pkg to 5.6.0 and using pkg-fetch 3.3 base binaries
Updated to TypeScript 4.5+
Updated @types/node to 16.11.7
Updated node-gyp-build to 4.4.0
Updated typedoc to 0.22.15, the .nojekyll generation is automatic
Changed to target ES2021 as node 16 supports it
Bin executable duct-tape in default.nix because node2nix no longer generates bin executables
Updated pkg builds hashes to v16.14.2 and specified target node range

emmacasolin committed:
build: updating core libraries

Updated js-async-init from 1.6.0 to 1.7.1
Updated js-async-locks from 2.0.0 to 2.2.0
Updated js-db from 3.2.0 to 3.2.3
Updating js-workers from 1.2.5 to 1.3.1

emmacasolin committed:
feat: static descriptions and exit codes for all errors

The `data` POJO that was originally supplied in the constructor as the second parameter is now part of the `options` parameter and most places in the code have been updated to match this.

#304

emmacasolin committed:
feat: using `cause` property to establish error chain

When any error is thrown as the result of another error occurring, the original error is now contained within the `cause` property of the new error.

#304

tegefaulkes committed:
tests: general fixes for failing tests

This contains fixes for failing tests as well as fixes for tests failing to exit when finished.

tegefaulkes committed:
syntax: added `@typescript-eslint/await-thenable` linting rule

This checks if we await things that are not promises. This is not a problem per se, but we generally don't want to await random things.

tegefaulkes committed:
fix: updated `@types/node-forge` version and fixed `keysUtils.getRandomBytes`

tegefaulkes committed:
tests: added test to check if nodes are properly added to the seed nodes when entering the network

This tests for if the Seed node contains the new nodes when they are created. It also checks if the new nodes discover each other after being created.

Includes a change to `findNode`. It will no longer throw an error when failing to find the node. This will have to be thrown by the caller now. This was required by `refreshBucket` since it's very likely that we can't find the random node it is looking for.

tegefaulkes committed:
tests: Added agent service tests for `nodesChainDataGet`, `nodesClosestLocalNode` and `nodesHolePunchMessage`

emmacasolin committed:
fix: validation errors are forced to use `void` as `cause`

#304

emmacasolin committed:
feat: error serialisation/deserialisation over gRPC incorporating error chaining

Our gRPC `toError` and `fromError` utils are now able to serialise and deserialise Polykey and non-Polykey errors (as well as non-errors), including the entire error chain if this exists. Also includes the ability to filter out sensitive data, for example when the error is being sent to another agent.

Errors sent over the network in this way are now additionally wrapped on the receiving side in an `ErrorPolykeyRemote` to make the source of the error more clear.

#304

emmacasolin committed:
feat: added logger to service handlers

All of the agent and client service handlers are now passed a shared logger that is used to log errors to stderr.

#304

emmacasolin committed:
syntax: removing index imports

emmacasolin committed:
fix: integrating db changes into ACL

emmacasolin committed:
fix: changing `es6` to `es2021` in `./eslintrc`

emmacasolin committed:
fix: integrating db changes into ACL

tegefaulkes committed:
fix: integrating db changes into vaults

tegefaulkes committed:
fix: integrating db changes into Notifications

Note that the `transaction` and `_transaction` methods still need te removed.

tegefaulkes committed:
fix: integrating db changes into `Discovery`

I opted not to use transactions for the most part. Adding things to the queue has to be handled with locking anyway so I opted to rely on that istead. This still bears testing though.

tegefaulkes committed:
fix: integrating db changes into `Sigchain`

emmacasolin committed:
build: updated dependencies and added incremental build

emmacasolin committed:
fix: integrating db changes into sessions

emmacasolin committed:
fix: keymanager tests passing

emmacasolin committed:
fix: git tests passing

emmacasolin committed:
fix: NodeGraph tests passing

emmacasolin committed:
feat: adding transactions to service handlers

tegefaulkes committed:
fix: fix for `GRPCServer` `http2Servers` hack

This was throwing an error because the expected object structure changed. The typing is brittle here since we're going around typescript's back to get a private property.

tegefaulkes committed:
fix: Small test fixes

tegefaulkes committed:
fix: Small fix for NotificationsManager tests

tegefaulkes committed:
fix: Small fix for Discovery tests

tegefaulkes committed:
fix: removing sigchain.transaction

emmacasolin committed:
fix: added locking back into sigchain

Sequence numbers and claim ids needed locking due to serialisation concerns

emmacasolin committed:
fix: added locking back into notifications

Notifications have strict ordering in the DB, therefore locking is required.

emmacasolin committed:
fix: discovery test fixes

emmacasolin committed:
fix: fix for `NodeConnection.test.ts` not running

emmacasolin committed:
refactor: replacing `Mutex` with `Lock` for Proxy

tegefaulkes committed:
fix: fixing vaultManager tests

There are still some tests failing due to cloning and pulling. Seems like a problem with the HTTP API for cloning.

tegefaulkes committed:
test: fixing client tests

tegefaulkes committed:
test: fixes for agent tests

tegefaulkes committed:
test: fixes for grpc tests

emmacasolin committed:
feat: error serialisation/deserialisation + metadata

tegefaulkes committed:
tests: fixing bin tests

Had to fix `expectProcessError` to work with the raw JSON forms of the errors rather than attempting to parse the errors. This is due to the bin CLI errors being separate from the normal errors and the reviver being unable to revive them.

I also had to fix how the `ErrorPolykeyRemote` converted the metadata with toJSON/fromJSON.

tegefaulkes committed:
feat: adding transactions to agent service handlers

emmacasolin committed:
fix: toError deserialization

There was a bug with the `fromJSON` method where it couldn't deserialise an error with an undefined cause (because this is removed by the replacer).

emmacasolin committed:
feat: adding metadata of sending node to errors

tegefaulkes committed:
feat: adding transactions to client service handlers

emmacasolin committed:
feat: NodeId in ErrorPolykeyRemote data now encoded

tegefaulkes committed:
feat: updated `NodeConnectionManager`'s resourceAcquire to handle errors properly

Making use of the updated API to use the error provided to the resourceRelease.

tegefaulkes committed:
feat: updated `NodeConnectionManager` to use `LockBox` for object map locking

Using `LockBox` allowed us to greatly simplify using the object map and creating connections.

tegefaulkes committed:
feat: updated `VaultManager` to use `LockBox` for object map locking

Using `LockBox` allowed us to greatly simplify using the object map and creating vaults.

emmacasolin committed:
feat: format json affects error logging

Specifying `--format=json` will cause errors to be logged in json format. ErrorPolykeyRemote errors also now contain additional metadata about the origin node of the error.

#323

tegefaulkes committed:
fix: added logic check to `NotificationsManager.readNotifications`

There was a bug with the DB that is getting fixed. This just throws a clear error when a notification can't be read despite it being in the list of all notifications.

emmacasolin committed:
tests: split vaults client service tests

tegefaulkes committed:
test: fixed broken proxy tests

Problem was with the tests themselves so i've fixed the tests. If there are any similar issues crop up later, then we can explore the underlying cause in depth. However, it is low priority for now.

CMCDragonkai committed:
tests: general clean up for proxy tests

tegefaulkes committed:
tests: fixing remaining tests

emmacasolin committed:
tests: grpc tests fixes

There were places where ClientMetadata was not being provided to grpc calls (`{} as ClientMetadata`) however this information is required in order to construct ErrorPolykeyRemote. Where metadata is needed in tests it is either now taken from constants used in the tests or mocked. There was also an issue with error descriptions not being serialised/deserialised which is now fixed. Also some brittle test cases for error messages were removed.

emmacasolin committed:
tests: split vaults secrets client service tests

emmacasolin committed:
tests: fix for nodesClaim test TypeError

tegefaulkes committed:
tests: fixed write after end error in proxy tests

tegefaulkes committed:
tests: fixed odd internal error for one test

It was just a problem with trying to connect to '0.0.0.0' with the UTP socket.

tegefaulkes committed:
syntax: general clean up proxy tests

tegefaulkes committed:
tests: fixing timeouts for tests failing in CI/CD

tegefaulkes committed:
feat: updating `@matrixai/db` to `^4.0.1`

Had to fix the usage in 5 domains.

tegefaulkes committed:
build: updated `@matrixai/db` and `encryptedfs` dependencies

tegefaulkes committed:
fix: updated usage of iterators to use `valueAsBuff: false` option

tegefaulkes committed:
fix: importing `LockRequest` for `VaultManager`

tegefaulkes committed:
fix: bug with `decryptWithKey` utility

tegefaulkes committed:
build: Updated `encryptedfs` and `@matrixai/db` version

emmacasolin committed:
refactor: moving `ClientMetadata` to GRPC domain
Having `ClientMetadata` in the top-level types caused an import loop. By moving it and `ErrorPolykeyRemote` to the grpc domain this loop is resolved. This was the best option since now all usage of `ClientMetadata` is inside the grpc domain.

emmacasolin committed:
fix: changed "remote error" to "cause" in cli error logging
#323

emmacasolin committed:
feat: filtering error logging in service handlers

We only want to log out server errors on the server side, so we now apply filters on errors before logging them.

tegefaulkes committed:
fix: post re-base fixes

DB and type changes using new transactions
Linting
Fixing timeouts

emmacasolin committed:
feat: updated `Opaque` usage

emmacasolin committed:
feat: `promisify` utility now preserves typing info

emmacasolin committed:
fix: bug fixes in vaults service handlers

emmacasolin committed:
style: lint fix

emmacasolin committed:
fix: service handlers check for client error chains

Sometimes an error is only considered a client error when part of the chain of another error (for example ErrorPolykeyRemote) so it is now possible to specify this array.

#304

emmacasolin committed:
ci: staging branch and CI/CD changes

tegefaulkes committed:
syntax: style fixes

Exports moved to end of the file as part of the `export {}` block
Cleaned up TODOs and FIXMEs
Fixed index and other incorrect imports
Type fixes in utils

tegefaulkes committed:
build: removing `node-abort-controller` polyfill

Abort controller functionality is included in node now.

tegefaulkes committed:
build: removing unneeded files

tegefaulkes committed:
fix: `NodeConnectionManager.syncNodeGraph` now pings nodes

Need to ensure validity of nodes by pinging them before adding them to the node graph.

#322

CMCDragonkai commented on package.json:
```diff
@@ -50,6 +50,7 @@
   "types": "dist/index.d.ts",
   "pkg": {
     "assets": [
+      "node_modules/jose/**/*",
```
We don't know why `jose` is needed. But you won't be able to tell until you actually run the final program. I recommend leaving it here until we verify why `jose` as an asset is required.

CMCDragonkai replied:
I can't find any native addons or JSON files except for `package.json` in jose.

CMCDragonkai replied:
This will require a manual check until we have integration testing for the built binaries.

tegefaulkes committed:
fix: small fixes to nodes utils parsing functions

emmacasolin committed:
test: `describeIf` and `testIf` utilities for conditional testing

tegefaulkes committed:
tests: expanding tests for root keypair changing

Added some tests to check that a root keyPair change propagates properly. Also added tests for the change for existing and new node connections.

#317

tegefaulkes committed:
build: updating package-lock.json

tegefaulkes committed:
fix: removed unneeded timer from NodeConnection test

CMCDragonkai committed:
nix: fixed building node2nix attribute in utils.nix

The package-lock.json was corrupted or not parseable by node2nix
After destroying the `node_modules` directory and deleting the
`package-lock.json`, we were able to recreate a working
`package-lock.json` file

If this occurs again, just destroy the `package-lock.json` and
`node_modules` directory and regenerate.

emmacasolin committed:
ci: using `--runInBand` for jest tests

emmacasolin committed:
tests: increasing timeouts for bin tests in CI/CD

emmacasolin committed:
tests: increasing timeouts for bin tests in CI/CD

emmacasolin committed:
fix: issue with ts-node causing increased test times in bin tests

`ts-node` needs to be fixed to 10.7.0 for now, as 10.8.1 causes many bin tests (those using `pkSpawn`) to timeout. Also reduced test timeouts back to what they originally were now that the timeout problem is resolved.

emmacasolin committed:
feat: better logging for service handler errors

If an error was logged out by a service handler it would previously appear as `createClientService:`, which was too vague. The logger will now state the name of the handler.

emmacasolin committed:
fix: PolykeyAgent now stops nodes domains when startup fails

emmacasolin committed:
fix: ICE protocol uses `Promise.any` instead of `Promise.all`

emmacasolin committed:
tests: fixes from testnet changes

The testnet PR brought in some changes that affect the NAT tests, so they needed to be modified slightly.
- Adding a node to the node graph pings the node by default. We want to disable this in the NAT tests so we have more control over the pings.
- Nodes add the details of any node that pings them. We can now remove some additional `nodes add` calls that were required to imitate this previously missing functionality.

tegefaulkes committed:
fix: expanded tests and small fix for `parseSeedNodes`

Related #324

emmacasolin committed:
build: upgraded `async-init`, `async-locks` and `js-logger`

emmacasolin committed:
feat: removing `edmSimple` configuration

Now that nodes add the details of a node that contacts them, we no longer need the `edmSimple` configuration to do this manually.

emmacasolin committed:
style: linting and other style fixes

General linting, using capitals for constants in NAT utils, and lowercase test descriptions

tegefaulkes committed:
fix: added `tslib` to `pkg` assets

This is to fix a problem where `tslib` was not getting included during the packaging process resulting in an error.

tegefaulkes committed:
build: fix for `build:macos` job

Related https://github.com/MatrixAI/js-async-locks/pull/10#issuecomment-1161934020

emmacasolin committed:
fix: relay node for hole punch message now re-writes proxy address

A relay node for a hole punch message was previously not modifying the proxy address in the message (which is the "return address" used to contact the source node). For nodes behind a NAT, who do not know their own public address, they rely on this overwriting so that nodes do not try to contact them on their private, inaccessible address.

tegefaulkes committed:
feat: added ability to pause `refreshBucketQueue`

This will allow us to disable the queue for testing.

emmacasolin committed:
build: package upgrades to aid testing

emmacasolin committed:
style: lintfix

emmacasolin committed:
feat: added diagrams to DMZ tests

#388

tegefaulkes committed:
fix: async bug with composing connections

the composed flag was set at the beginning of the compose function causing another function to throw an error due to an undefined property if it was called at the same time.

emmacasolin committed:
Merge pull request #381 from MatrixAI/feature-nat-testing

Tests for NAT-Traversal and Hole-Punching

tegefaulkes committed:
build: fix for `integration:nix` CI/CD job

Problem was to do with the `nix-build` process not patching the shebangs of the `polykey.js` script. this resulted in it failing to run due to not finding node in it's environment. The fix was to make `src/bin/polykey.js` and  `src/bin/polykey-agent.ts` executable.
